### ISSUE TITLE
[Registration] Incomplete tests and possible bug

### ISSUE THREAD

**Issue by mvieth** [PointCloudLibrary/pcl#4455]
In the files `test_ndt.cpp`, `test_registration.cpp`, and `test_sac_ia.cpp`, there are bad bool casts (`(bool) iter/2` and `(bool) iter%2`, 12 instances total). Since the cast has a higher precedence (executed first) than the division/modulo operators, only two of the actually intended four test cases are executed (see also here: https://godbolt.org/z/EKx5x5). So either parentheses should be added (`(bool) (iter/2)`) or the cast completely eliminated (`(iter/2)!=0`).

When I did that, segmentation faults appeared. Valgrind says they happen [here](https://github.com/PointCloudLibrary/pcl/blob/9c592cb116ecb67f773a630fe3219687c91ab8bb/common/include/pcl/point_representation.h#L151), when `vectorize` is called from [here](https://github.com/PointCloudLibrary/pcl/blob/9c592cb116ecb67f773a630fe3219687c91ab8bb/kdtree/include/pcl/kdtree/impl/kdtree_flann.hpp#L145), because `dim_` is zero. `dim_` is initially zero and set to the correct value [here](https://github.com/PointCloudLibrary/pcl/blob/9c592cb116ecb67f773a630fe3219687c91ab8bb/kdtree/include/pcl/kdtree/impl/kdtree_flann.hpp#L97), probably through [here](https://github.com/PointCloudLibrary/pcl/blob/9c592cb116ecb67f773a630fe3219687c91ab8bb/kdtree/include/pcl/kdtree/kdtree.h#L113) and [here](https://github.com/PointCloudLibrary/pcl/blob/9c592cb116ecb67f773a630fe3219687c91ab8bb/registration/include/pcl/registration/impl/registration.hpp#L202).

I haven't used the registration module much yet, so I would like some help here. To me, [this if-statement](https://github.com/PointCloudLibrary/pcl/blob/9c592cb116ecb67f773a630fe3219687c91ab8bb/registration/include/pcl/registration/registration.h#L219) (and further down in `setSearchMethodSource`) looks suspicious, and might cause the segfault. Why not just do `force_no_recompute_ = force_no_recompute`? The if-statement means that if `force_no_recompute_` is set to `true` once, it stays that way. @koide3 Maybe you can shed some light on this one?

**Comment by koide3** [comment]
@mvieth 
I agree with you that the if-statement should be the cause. When ```force_no_recompute_ == true```, ```pcl::Registration``` doesn't call ```tree->setInputCloud()``` inside, and the user must explicitly call it by themselves. The test code doesn't call ```tree->setInputCloud``` when ```force_cache == false```, but the bug makes ```force_no_recompute_``` stay being ```true```, then a tree with an empty cloud can be used and cause a segfault.

I believe ```setSearchMethodTarget()``` should just do ```force_no_recompute_ = force_no_recompute```, but I'm not 100% sure how this fix changes the behavior of other codes. I think we need to check if there is no code relying on the strange behavior of ```force_no_recompute_```. Note that it seems ```correspondence_estimation.h```, ```correspondence_rejection.h```, and ```transformation_validation_euclidean.h``` have the same bug.

**Comment by JStech** [comment]
I grepped the library for calls to `setSearchMethod(Target|Source)` and I only found the `force_no_recompute` param specified in tests and within the registration module. Within the module, it's used in a few functions in `correspondence_rejection_(var_trimmed|median_distance|surface_normal|distance).h` like this:
```
  template <typename PointT>
  inline void
  setSearchMethodTarget(const typename pcl::search::KdTree<PointT>::Ptr& tree,
                        bool force_no_recompute = false)
  {
    static_pointer_cast<DataContainer<PointT>>(data_container_)
        ->setSearchMethodTarget(tree, force_no_recompute);
  }
```
Otherwise, it only appears in `registration.h` and `impl/registration.hpp` within the module. Some of the tests do set the option, which I want to look at a bit more closely. It looks like nothing in the library depends on this strange behavior.
